exp/internal/controllers/machinepool_controller.go

```diff
@@ -253,33 +253,52 @@ func (r *MachinePoolReconciler) reconcile(ctx context.Context, cluster *clusterv
 	return res, kerrors.NewAggregate(errs)
 }
 
-func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, mp *expv1.MachinePool) error {
-	if ok, err := r.reconcileDeleteExternal(ctx, mp); !ok || err != nil {
-		// Return early and don't remove the finalizer if we got an error or
-		// the external reconciliation deletion isn't ready.
-		return err
-	}
+func (r *MachinePoolReconciler) reconcileDelete(ctx context.Context, cluster *clusterv1.Cluster, machinepool *expv1.MachinePool) error {
+	deleteAllowed, clusterClient, err := r.isDeleteMachinePoolAllowed(ctx, cluster)
+
+	// Check for cluster allowing delete or machinePool delete timeout.
+	if deleteAllowed || r.isMachinePoolDeleteTimeoutPassed(machinepool) {
+		if ok, err := r.reconcileDeleteExternal(ctx, machinepool); !ok || err != nil {
+			// Return early and don't remove the finalizer if we got an error or
+			// the external reconciliation deletion isn't ready.
+			return err
+		}
 
-	if err := r.reconcileDeleteNodes(ctx, cluster, mp); err != nil {
-		// Return early and don't remove the finalizer if we got an error.
-		return err
+		// Delete nodes when cluster accessor available & there are nodes to delete.
```

Comment:
should this PR be scoped only to retrieve the cluster client and skip if not functional?